src/env/ParticleGrid.cpp

```diff
@@ -0,0 +1,240 @@
+
+                    for (const auto d : displacements) {
+                        int3 idx2 = idx1 + d;
+                        if (idx1 >= idx2  // ensure only unique pairs are added
+                            || idx2[0] < 0 || idx2[1] < 0 || idx2[2] < 0 ||
+                            static_cast<UINT_T>(idx2[0]) >= cell_count[0] ||
+                            static_cast<UINT_T>(idx2[1]) >= cell_count[1] ||
+                            static_cast<UINT_T>(idx2[2]) >= cell_count[2]) {
+                            continue;
+                        }
+
+                        cell_pairs.emplace_back(cells.at(idx1), cells.at(idx2));
+                    }
+                }
+            }
+        }
+    }
+
+    /// -----------------------------------------
+    /// \brief interaction with ParticleGrid
+    /// -----------------------------------------
+    const GridCell& ParticleGrid::get_cell(const int3& idx) const { return cells.at(idx); }
+
+    const GridCell& ParticleGrid::get_cell(const Particle& particle) const {
+        return cells.at(what_cell(particle.position));
+    }
+
+    GridCell& ParticleGrid::get_cell(const int3& idx) { return cells.at(idx); }
+
+    GridCell& ParticleGrid::get_cell(const Particle& particle) { return cells.at(what_cell(particle.position)); }
+
+    int3 ParticleGrid::what_cell(const vec3& position) const {
+        const vec3 pos = position_in_grid(position);
+        // boundary is axis aligned
+        if (pos[0] < 0 || pos[1] < 0 || pos[2] < 0) {
+            return OUTSIDE_CELL;
+        }
+
+        const auto x = static_cast<INT_T>(pos[0] / cell_size[0]);
+        const auto y = static_cast<INT_T>(pos[1] / cell_size[1]);
+        const auto z = static_cast<INT_T>(pos[2] / cell_size[2]);
+
+        if (static_cast<UINT_T>(x) >= cell_count[0] || static_cast<UINT_T>(y) >= cell_count[1] ||
+            static_cast<UINT_T>(z) >= cell_count[2]) {
+            return OUTSIDE_CELL;
+        }
+
+        return int3{x, y, z};
+    }
+
+    std::vector<int3> ParticleGrid::get_cell_indices() const {
+        std::vector<int3> keys;
+        for (const auto& key : cells | std::views::keys) {
+            keys.push_back(key);
+        }
+        return keys;
+    }
+
+    const std::vector<GridCellPair>& ParticleGrid::linked_cells() {
+        return cell_pairs;
+    }
+
+    const std::vector<GridCell*> & ParticleGrid::boundary_cells() {
+        return border_cells;
+    }
+
+    // std::vector<GridCell> ParticleGrid::grid_cells() {
+    //     // not the best design but this function is intended for debugging purposes so its ok
+    //     std::vector<GridCell> cells;
+    //     cells.reserve(this->cells.size());
+    //
+    //     for (auto& snd : this->cells | std::views::values) {
+    //         cells.push_back(snd);
+    //     }
+    //
+    //     return cells;
+    // }
+
+    void ParticleGrid::update_cells(Particle* particle, const int3& old_cell, const int3& new_cell) {
```

Comment:
i am aware of smart pointers, however their use is (AFAIK) usually associated with RAII. Here the ownership and lifetime are clear: they are directly coupled with the vector particle_storage in Environment. Hence i do not see any benefit in using a smart pointer since we are not transferring or sharing any ownership.

Reply:
Yes, reference would have been more cleaner in my opinion. Also, this has less to do with functionality and more to do with good coding practices (using references, smart pointers, STL functions, etc.). In any case, this was just a recommendation and there is nothing wrong here :)